middleware/markdown/markdown.go

```diff
@@ -99,74 +70,67 @@ func (md Markdown) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
+			if err != nil {
+				return http.StatusInternalServerError, err
+			}
+			dirents, err = fdp.Readdir(-1)
+			if err != nil {
+				return http.StatusInternalServerError, err
+			}
+
+			// Set path to found index file
 			fpath = idx
 		}
 
-		for _, ext := range cfg.Extensions {
-			if strings.HasSuffix(fpath, ext) {
-				f, err := md.FileSys.Open(fpath)
-				if err != nil {
-					if os.IsPermission(err) {
-						return http.StatusForbidden, err
-					}
-					return http.StatusNotFound, nil
-				}
-
-				fs, err := f.Stat()
-				if err != nil {
-					return http.StatusNotFound, nil
+		// If supported extension, process it
+		if _, ok := cfg.Extensions[path.Ext(fpath)]; ok {
+			f, err := md.FileSys.Open(fpath)
+			if err != nil {
+				if os.IsPermission(err) {
+					return http.StatusForbidden, err
 				}
+				return http.StatusNotFound, nil
+			}
 
-				// if development is set, scan directory for file changes for links.
-				if cfg.Development {
-					if err := GenerateStatic(md, cfg); err != nil {
-						log.Printf("[ERROR] markdown: on-demand site generation error: %v", err)
-					}
-				}
+			fs, err := f.Stat()
+			if err != nil {
+				return http.StatusNotFound, nil
+			}
 
-				cfg.RLock()
-				filepath, ok := cfg.StaticFiles[fpath]
-				cfg.RUnlock()
-				// if static site is generated, attempt to use it
-				if ok {
-					if fs1, err := os.Stat(filepath); err == nil {
-						// if markdown has not been modified since static page
-						// generation, serve the static page
-						if fs.ModTime().Before(fs1.ModTime()) {
-							if html, err := ioutil.ReadFile(filepath); err == nil {
-								middleware.SetLastModifiedHeader(w, fs1.ModTime())
-								w.Write(html)
-								return http.StatusOK, nil
-							}
-							if os.IsPermission(err) {
-								return http.StatusForbidden, err
-							}
-							return http.StatusNotFound, nil
-						}
-					}
-				}
+			body, err := ioutil.ReadAll(f)
+			if err != nil {
+				return http.StatusInternalServerError, err
+			}
 
-				body, err := ioutil.ReadAll(f)
-				if err != nil {
-					return http.StatusInternalServerError, err
-				}
+			ctx := middleware.Context{
+				Root: md.FileSys,
+				Req:  r,
+				URL:  r.URL,
+			}
+			html, err := cfg.Markdown(fpath, body, dirents, ctx)
+			if err != nil {
+				return http.StatusInternalServerError, err
+			}
 
-				ctx := middleware.Context{
-					Root: md.FileSys,
-					Req:  r,
-					URL:  r.URL,
-				}
-				html, err := md.Process(cfg, fpath, body, ctx)
-				if err != nil {
-					return http.StatusInternalServerError, err
-				}
+			// TODO(weingart): move template execution here, something like:
+			//
+			// html, err = md.execTemplate(cfg, html, ctx)
+			// if err != nil {
+			// 	return http.StatusInternalServerError, err
+			// }
 
-				middleware.SetLastModifiedHeader(w, fs.ModTime())
-				w.Write(html)
-				return http.StatusOK, nil
-			}
+			middleware.SetLastModifiedHeader(w, fs.ModTime())
+			w.Write(html)
```

Comment:
I really wish that every middleware could always write, even on error.  Then, the middleware invoking stack would buffer this (bytes.Buffer), and do the "right thing" depending on verb (GET/HEAD), errors returned, etc.

It would make a lot of the middleware a lot simpler, and likely cause less copying and spurious buffers to be allocated.

For example, most of the template (and json, etc) things need to render a temporary buffer locally, and then once they've seen no errors, they can write as a last resort.

If instead the middleware invoking layer does the buffering, a lot of that logic and allocation goes away.  One downside, for serving static files, which may be sent with sendfile(2) today, there would need to be something figured out.  Possibly by having middleware specify the type of buffering they wish to have (default to buffered), such that static file sending could still use the optimized sendfile(2) system calls.